compiler/src/iree/compiler/Codegen/Common/ResolveSwizzleHints.cpp

```diff
@@ -0,0 +1,196 @@
+/// vector.store %2, %src[swizzleOffset(%src, %offset + 8)] : vector<4>
+/// vector.store %3, %src[swizzleOffset(%src, %offset + 12)] : vector<4>
+void swizzleStore(RewriterBase &rewriter, vector::StoreOp store,
+                  IREE::Codegen::SwizzleHintOp hintOp) {
+  Location hintLoc = hintOp.getLoc();
+  int64_t accessWidth = hintOp.getSwizzle().getAccessElementCount();
+  VectorType type = store.getVectorType();
+  int64_t storeWidth = type.getShape()[0];
+  Value memrefOffset = store.getIndices()[0];
+
+  AffineExpr s0, s1;
+  bindSymbols(rewriter.getContext(), s0, s1);
+  AffineMap sum = AffineMap::get(0, 2, s0 + s1);
+
+  // Store type = vector<C>, k = accessWidth
+  // i = 0 -> C += k is the offset into the vector of a contiguous group of
+  // swizzled elements.
+  for (int64_t i = 0; i < storeWidth; i += accessWidth) {
+    Value subVec = rewriter.create<vector::ExtractStridedSliceOp>(
+        store.getLoc(), store.getValueToStore(), ArrayRef<int64_t>{i},
+        ArrayRef<int64_t>{accessWidth}, ArrayRef<int64_t>{1});
+    auto vecOffset = rewriter.getIndexAttr(i);
+    auto newBaseOffset = affine::makeComposedFoldedAffineApply(
+        rewriter, hintLoc, sum, {memrefOffset, vecOffset});
+
+    Value newOffset = getValueOrCreateConstantIndexOp(
+        rewriter, hintLoc,
+        hintOp.getSwizzle().swizzleOffset(rewriter, hintOp.getLoc(),
+                                          newBaseOffset, hintOp.getOperand()));
+    rewriter.create<vector::StoreOp>(store.getLoc(), subVec, store.getBase(),
+                                     newOffset);
+  }
+  rewriter.eraseOp(store);
+}
+
+/// Resolves all hints. Walks all direct users and splits them into loads and
+/// stores. If any user is not a swizzle-able load or store, bail out and
+/// silently drop the optimization hint.
+void resolveHintOp(RewriterBase &rewriter,
+                   IREE::Codegen::SwizzleHintOp hintOp) {
+  SmallVector<vector::LoadOp> loads;
+  SmallVector<vector::StoreOp> stores;
+  int64_t accessWidth = hintOp.getSwizzle().getAccessElementCount();
+  for (Operation *user : hintOp->getUsers()) {
+    if (auto load = dyn_cast<vector::LoadOp>(user)) {
+      VectorType loadType = load.getVectorType();
+      // Guard on zero rank loads and loads not divisible by the access width.
+      if (loadType.getRank() != 1 ||
+          loadType.getShape()[0] % accessWidth != 0) {
+        return;
+      }
+      loads.push_back(load);
+      continue;
+    }
+    if (auto store = dyn_cast<vector::StoreOp>(user)) {
+      VectorType storeType = store.getVectorType();
+      // Guard on zero rank stores and stores not divisible by the access width.
+      if (storeType.getRank() != 1 ||
+          storeType.getShape()[0] % accessWidth != 0) {
+        return;
+      }
+      stores.push_back(store);
+      continue;
+    }
+    // Bail out if we can't rewrite all users.
+    return;
+  }
+
+  for (auto load : loads) {
+    rewriter.setInsertionPoint(load);
```

Comment:
Alternatively, you could make these helpers set the insertion point they need instead of assuming the caller does that. I don't think it matters too much either way.

Reply:
I prefer keeping insertion point changes outside of helpers because it's closer to the style of pattern rewrites. I don't mind changing it either way though.